### PR TITLE
feat: coalesce multiple reads together and don't block on io if there's values available

### PR DESCRIPTION
**Pull request by robert3005**

### CODE REVIEW

File: bench-vortex/src/bin/tpch_benchmark.rs [hunk 1]
```diff
@@ -135,7 +135,7 @@ fn render_table(receiver: Receiver<Measurement>, formats: &[Format]) -> anyhow::
 
     for (query_idx, baseline_measure) in baseline.iter().enumerate().take(22) {
         let query_baseline = baseline_measure.time.as_micros();
-        let mut row = vec![query_idx.to_string()];
+        let mut row = vec![(query_idx + 1).to_string()];
```
Comment (robert3005) [inline bench-vortex/src/bin/tpch_benchmark.rs]: split to separate pr

File: vortex-array/src/compute/filter.rs [hunk 2]
```diff
@@ -91,7 +91,6 @@ pub fn filter(array: &ArrayData, mask: FilterMask) -> VortexResult<ArrayData> {
 pub struct FilterMask {
     array: ArrayData,
     true_count: usize,
-    true_range: (usize, usize),
```
Comment (robert3005) [inline vortex-array/src/compute/filter.rs]: split to separate pr

File: vortex-file/src/read/coalescer.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
+#![allow(dead_code)]
```
Comment (robert3005) [inline vortex-file/src/read/coalescer.rs]: remove

File: vortex-file/src/read/coalescer.rs [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        self.layout.read_selection(mask)
+    }
+}
+
+enum RowMaskState<V> {
+    Pending(RowMask),
+    Ready(V),
+}
+
+pub struct Coalescer<R, S, V, RM> {
```
Comment (robert3005) [inline vortex-file/src/read/coalescer.rs]: Majority of interesting logic in this pr is this object. It performs multiple reads on underlying layout readers and dispatches them together. It makes sure there's ALWAYS one io request in flight. Importantly if the consumer of the stream is slow this might lead to using a lot of memory

File: vortex-file/src/read/coalescer.rs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+                                *queued_res = RowMaskState::Ready(v);
+                            }
+                        }
+                    }
+                }
+                RowMaskState::Ready(_) => {}
+            }
+        }
+
+        while read_more_count < NUM_TO_COALESCE {
```
Comment (robert3005) [inline vortex-file/src/read/coalescer.rs]: We are using arbitrary number of read requests to decide how much to coalesce

File: vortex-io/src/read_ranges.rs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    pub fn new(read: R, dispatcher: Arc<IoDispatcher>, max_gap: usize) -> VortexReadRanges<R> {
+        Self {
+            read,
+            dispatcher,
+            max_gap,
+        }
+    }
+}
+
+impl<R: VortexReadAt> VortexReadRanges<R> {
```
Comment (robert3005) [inline vortex-io/src/read_ranges.rs:33]: This is closely based on arrow-rs logic but we dispatch each request on our io loop regardless of underlying reader. We shold decide if we require ranges to be sorted by start or will the reader sort them (current implementation)

File: vortex-io/src/read_ranges.rs [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+#[derive(Debug, Clone)]
+pub struct VortexReadRanges<R> {
+    read: R,
+    dispatcher: Arc<IoDispatcher>,
+    max_gap: usize,
+}
+
+impl<R> VortexReadRanges<R> {
+    pub fn new(read: R, dispatcher: Arc<IoDispatcher>, max_gap: usize) -> VortexReadRanges<R> {
```
Comment (robert3005) [inline vortex-io/src/read_ranges.rs:24]: We should infer max_gap between ranges based on observed latency. This is just the simplest implementation

File: vortex-file/src/read/stream.rs [hunk 8]
```diff
@@ -50,22 +47,40 @@ impl<R: VortexReadAt> VortexFileArrayStream<R> {
         row_count: u64,
         row_mask: Option<RowMask>,
         dispatcher: Arc<IoDispatcher>,
-    ) -> Self {
+    ) -> VortexResult<Self> {
+        let mut reader_splits = BTreeSet::new();
```
Comment (robert3005) [inline vortex-file/src/read/stream.rs:51]: since we can construct all the readers eagerly now it's simplest to register all splits on creation

**Comment by robert3005** [comment]
Seems like ci hits a deadlock

**Comment by robert3005** [comment]
I must have introduced deadlock in https://github.com/spiraldb/vortex/pull/1466/commits/4ffd42bbcd049af04175208c4b285efc83be2584 since everything was running before that commit. I will figure it out tonight

File: vortex-file/src/read/coalescer.rs [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+            if !messages.is_empty() {
+                self.in_flight = Some(self.dispatch_messages(messages, cx.waker().clone()));
+            }
+        }
+
+        if let Some(next_ready) = self.queued.pop_front() {
+            match next_ready {
+                RowMaskState::Pending(m) => {
+                    self.queued.push_front(RowMaskState::Pending(m));
+                    Poll::Pending
```
Comment (robert3005) [inline vortex-file/src/read/coalescer.rs]: yes, this pending needs some way to know how to resume, I think reads should follow and this would resume as ready but I don't think it does

File: vortex-file/src/read/coalescer.rs [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    values_iter: S,
+    in_flight: Option<BoxFuture<'static, io::Result<Vec<Message>>>>,
+    queued: VecDeque<RowMaskState<V>>,
+    io_read: VortexReadRanges<R>,
+    dispatcher: Arc<IoDispatcher>,
+    row_mask_reader: RM,
+    cache: Arc<RwLock<LayoutMessageCache>>,
+}
+
+impl<R, S, V, RM> Coalescer<R, S, V, RM>
```
Comment (a10y) [inline vortex-file/src/read/coalescer.rs]: name is a bit off since the coalescing is actually happening in the VortexReadRanges

File: vortex-file/src/read/coalescer.rs [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+        dispatcher: Arc<IoDispatcher>,
+        values_iter: S,
+        row_mask_reader: RM,
+        cache: Arc<RwLock<LayoutMessageCache>>,
+    ) -> Self {
+        Self {
+            values_iter,
+            in_flight: None,
+            queued: VecDeque::new(),
+            io_read: VortexReadRanges::new(read, dispatcher.clone(), 1 << 20),
```
Comment (lwwmanning) [inline vortex-file/src/read/coalescer.rs]: lol @ magic `1<<20` (presumably 1MiB?)

File: vortex-io/src/read_ranges.rs [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+                .checked_sub(range_end)
+                .map(|gap| gap <= max_gap)
+                .unwrap_or(true)
+        {
+            range_end = range_end.max(ranges[end_idx].end);
+            end_idx += 1;
+        }
+
+        merged_ranges.push(ranges[start_idx].start..range_end);
+        start_idx = end_idx;
```
Comment (lwwmanning) [inline vortex-io/src/read_ranges.rs:107]: end_idx is inclusive, so I think this should be `start_idx = end_idx + 1`

File: vortex-file/src/read/coalescer.rs [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+    type Item = VortexResult<V>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if let Some(in_flight) = &mut self.in_flight {
+            match in_flight.poll_unpin(cx) {
+                Poll::Ready(msgs) => {
+                    self.store_messages(
+                        msgs.map_err(|e| vortex_err!("Cancelled in flight read {e}"))?,
+                    );
+                    let messages = ready!(self.gather_read_messages(cx))?;
```
Comment (robert3005) [inline vortex-file/src/read/coalescer.rs]: If this returns pending we get into an error state where we poll a done future which in rust is a bug. Need this to not propagate pending

File: vortex-array/src/compute/filter.rs [hunk 14]
```diff
@@ -202,6 +204,10 @@ impl FilterMask {
         self.array.len() - self.true_count
     }
 
+    pub fn true_range(&self) -> (usize, usize) {
```
Comment (lwwmanning) [inline vortex-array/src/compute/filter.rs]: we got rid of true_range in #1468, prob want to remove here as well

Comment (robert3005) [reply vortex-array/src/compute/filter.rs]: Yup, I’m working on fixing the last issue today and have it rebased locally

File: vortex-file/src/read/metadata.rs [hunk 15]
```diff
@@ -33,74 +51,24 @@ impl<R: VortexReadAt + Unpin> MetadataFetcher<R> {
...
-        Self {
+        let metadata_reader = Coalescer::new(
             input,
             dispatcher,
-            root_layout,
+            Box::new(stream::iter(iter::once(Ok(RowMask::new_valid_between(
+                0, 1,
+            ))))) as _,
+            MetadataMaskReader::new(root_layout),
             layout_cache,
```
Comment (robert3005) [inline vortex-file/src/read/metadata.rs]: This suggests broken abstraction. We require some initial value always and we produce this empty range

Comment (robert3005) [reply vortex-file/src/read/metadata.rs]: However, it's just simpler to delegate to the existing logic and not have another type. We should factor out common logic from Coalesce for single element operation

File: vortex-file/src/read/mod.rs [hunk 16]
```diff
@@ -57,20 +59,18 @@ pub type MessageId = Vec<LayoutPartId>;
...
+pub type BatchRead = MessageRead<ArrayData>;
+pub type SplitRead = MessageRead<RowMask>;
+pub type MetadataRead = MessageRead<Vec<Option<ArrayData>>>;
 
 #[derive(Debug)]
-pub enum MetadataRead {
-    /// Layout has no metadata
-    None,
-    /// Additional IO is required
+pub enum MessageRead<T> {
```
Comment (robert3005) [inline vortex-file/src/read/mod.rs:72]: Add `None` as a variant and change and remove nested value from Option in all return types

File: vortex-io/src/read_ranges.rs (hunk 12 above)
Comment (robert3005) [reply vortex-io/src/read_ranges.rs:107]: I think you're right. This is repeating the last chunk read twice

Comment (robert3005) [reply vortex-io/src/read_ranges.rs:107]: No, the previous loop stops only if the gap is bigger so once the gap is bigger than the max gap then end_idx is one further than we'd like to merge so it's the next range to consider

File: vortex-file/src/read/builder/mod.rs [hunk 17]
```diff
@@ -71,7 +71,7 @@ pub struct VortexReadBuilder<R> {
     io_dispatcher: Option<Arc<IoDispatcher>>,
 }
 
-impl<R: VortexReadAt> VortexReadBuilder<R> {
+impl<R: VortexReadAt + Unpin> VortexReadBuilder<R> {
```
Comment (robert3005) [inline vortex-file/src/read/builder/mod.rs:74]: We shouldn't be so `Unpin` and instead use `pin_project` more

File: vortex-file/src/read/coalescer.rs (hunk 10 above)
Comment (robert3005) [reply vortex-file/src/read/coalescer.rs]: you're right, this is the thing that buffers values so they can be coalesced

File: vortex-file/src/read/coalescer.rs (hunk 11 above)
Comment (robert3005) [reply vortex-file/src/read/coalescer.rs]: yes, we should get rid of it and instead messure the value in the reader

File: vortex-file/src/read/splits.rs [hunk 18]
```diff
@@ -1,61 +1,43 @@
...
-/// Iterator over row ranges of a vortex file with bitmaps of valid values in those ranges
-pub trait MaskIterator: Iterator<Item = VortexResult<SplitMask>> + Send {
-    /// Register additional horizontal row boundaries to split the generated layout on
-    fn additional_splits(&mut self, splits: &mut BTreeSet<usize>) -> VortexResult<()>;
-}
-
-enum PruningSplitState {
-    Ranges(VecDeque<(usize, usize)>),
-    Splits(BTreeSet<usize>),
+impl RowMaskLayoutReader {
```
Comment (a10y) [inline vortex-file/src/read/splits.rs]: - RowMaskLayoutReader
- MaskLayoutReader
- RowMaskReader

Are a confusing named trio

File: vortex-file/src/read/buffered.rs [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+        Ok((to_read, exhausted))
+    }
+
+    fn dispatch_messages(
+        &self,
+        messages: Vec<MessageLocator>,
+        waker: Waker,
+    ) -> BoxFuture<'static, io::Result<Vec<Message>>> {
+        let reader = self.io_read.clone();
+        self.dispatcher
```
Comment (a10y) [inline vortex-file/src/read/buffered.rs:165]: I don't think you need this outer `dispatcher.dispatch`, since `read_byte_ranges` is also going through the Dispatcher.

File: vortex-io/src/read_ranges.rs [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+            max_gap,
+        }
+    }
+}
+
+impl<R: VortexReadAt> VortexReadRanges<R> {
+    pub fn read_byte_ranges(
+        &self,
+        ranges: Vec<Range<usize>>,
+    ) -> impl Future<Output = io::Result<Vec<Bytes>>> + 'static {
```
Comment (a10y) [inline vortex-io/src/read_ranges.rs]: We should probably update the signature to reflect that the returned future is `Send`, or perhaps just return a `BoxFuture` directly

File: vortex-file/src/read/metadata.rs [hunk 21]
```diff
@@ -33,74 +51,22 @@ impl<R: VortexReadAt + Unpin> MetadataFetcher<R> {
         root_layout: Box<dyn LayoutReader>,
         layout_cache: Arc<RwLock<LayoutMessageCache>>,
     ) -> Self {
-        Self {
+        let metadata_reader = BufferedLayoutReader::new(
             input,
             dispatcher,
-            root_layout,
+            stream::iter(iter::once(Ok(RowMask::new_valid_between(0, 1)))),
```
Comment (a10y) [inline vortex-file/src/read/metadata.rs:59]: why [0,1]

File: vortex-file/src/read/buffered.rs [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+        Self { layout }
+    }
+}
+
+impl RowMaskReader<ArrayData> for MaskLayoutReader {
+    /// Read given mask out of the reader
+    fn read_mask(&self, mask: &RowMask) -> VortexResult<Option<MessageRead<ArrayData>>> {
+        self.layout.read_selection(mask)
+    }
+}
```
Comment (a10y) [inline vortex-file/src/read/buffered.rs]: I think you can kill this and just replace it with a blanket impl

```
impl<T: LayoutReader> RowMaskReader<ArrayData> for T {
    fn read_mask(&self, mask: &RowMask) -> VortexResult<Option<MessageRead<ArrayData>>> {
        self.read_selection(mask)
    }
}
```

File: vortex-file/src/read/mod.rs [hunk 23]
```diff
@@ -57,28 +59,19 @@ pub type MessageId = Vec<LayoutPartId>;
...
-
-#[derive(Debug)]
-pub enum MetadataRead {
-    /// Layout has no metadata
-    None,
-    /// Additional IO is required
-    ReadMore(Vec<MessageLocator>),
-    Batches(Vec<Option<ArrayData>>),
-}
+pub type BatchRead = MessageRead<ArrayData>;
```
Comment (a10y) [inline vortex-file/src/read/mod.rs:66]: I personally find these type aliases more confusing than using the underlying type, though I think that is controversial

File: vortex-file/src/read/metadata.rs (hunk 21 above)
Comment (robert3005) [reply vortex-file/src/read/metadata.rs:59]: This is bad abstraction. This rowmask is never used but I need to produce a stream of row masks to call the metadata read… need to think this one over

File: vortex-file/src/read/buffered.rs (hunk 22 above)
Comment (a10y) [reply vortex-file/src/read/buffered.rs]: nvm

File: vortex-file/src/read/buffered.rs (hunk 19 above)
Comment (robert3005) [reply vortex-file/src/read/buffered.rs:165]: I had a deadlock without it. The await is never resumed as we propagate the pending to the stream but if stream returns pending it needs to wake itself up so I need something to poll the read and then call wake

File: vortex-file/src/read/buffered.rs [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+use futures_util::{FutureExt, StreamExt};
+use vortex_array::ArrayData;
+use vortex_error::{vortex_err, vortex_panic, VortexExpect, VortexResult};
+use vortex_io::{Dispatch, IoDispatcher, VortexReadAt, VortexReadRanges};
+
+use crate::{LayoutMessageCache, LayoutReader, Message, MessageLocator, MessageRead, RowMask};
+
+const NUM_TO_COALESCE: usize = 8;
+
+pub trait RowMaskReader<V> {
```
Comment (a10y) [inline vortex-file/src/read/buffered.rs]: all of the types in this file should probably either be non-`pub` or at most `pub(crate)` to avoid leaking impl things into public API, with some more imminent usage

**Review by a10y — APPROVED** [review]
I made a PR to clean up some naming and add some extra comments here: https://github.com/spiraldb/vortex/pull/1520

I think the structure of the reader has gotten really complex, with a lot of mutable state, lots of `.expect` I'd like to remove, some awkward names. I think we can work on that after this merges though so we capture the perf improvements

**Comment by robert3005** [comment]
I merged the renames. I think there's still something funky with `Unpin` but I will fix it in a followup
